faculty/session.py

```diff
@@ -26,20 +30,116 @@
 AccessToken = namedtuple("AccessToken", ["token", "expires_at"])
 
 
+class AccessTokenSchema(Schema):
+    token = fields.String(required=True)
+    expires_at = fields.DateTime(data_key="expiresAt", required=True)
+
+    @post_load
+    def make_access_token(self, data):
+        return AccessToken(**data)
+
+
+class AccessTokenStore:
+    def __init__(self, tokens=None):
+        self.tokens = tokens or {}
+
+    @staticmethod
+    def _hash_profile(profile):
+        return str(hash(profile))
+
+    def __getitem__(self, profile):
+        return self.tokens[self._hash_profile(profile)]
+
+    def __setitem__(self, profile, access_token):
+        self.tokens[self._hash_profile(profile)] = access_token
+
+    def get(self, profile):
+        try:
+            return self[profile]
+        except KeyError:
+            return None
+
+
+class AccessTokenStoreSchema(Schema):
+    tokens = fields.Dict(
+        keys=fields.String(),
+        values=fields.Nested(AccessTokenSchema),
+        required=True,
+    )
+
+    @post_load
+    def make_access_token_store(self, data):
+        return AccessTokenStore(**data)
+
+
+def _is_valid_access_token(access_token_or_none):
+    if access_token_or_none is None:
+        return False
+    else:
+        return access_token_or_none.expires_at >= datetime.now(tz=pytz.utc)
+
+
 class MemoryAccessTokenCache(object):
     def __init__(self):
-        self._store = {}
+        self._store = AccessTokenStore()
 
     def get(self, profile):
         access_token = self._store.get(profile)
-        utc_now = datetime.now(tz=pytz.utc)
-        if access_token is None or access_token.expires_at < utc_now:
-            return None
+        return access_token if _is_valid_access_token(access_token) else None
+
+    def add(self, profile, access_token):
+        self._store[profile] = access_token
+
+
+def _ensure_directory_exists(path, mode):
+    try:
+        os.makedirs(path, mode=mode)
+    except OSError as e:
+        if e.errno == errno.EEXIST:
+            # Directory already exists
```

Comment:
I'm not sure this comment is necessary, given the use of a named constant in the comparison to `e.errno`.

Reply:
True, though we use `errno.ENOENT` elsewhere which is much less clear and I liked being consistent and having the comment in both places. What do you think?

Reply:
Either way, consistency is good too.